pkg/engine/engine.go

```diff
@@ -134,270 +144,64 @@ func (e *Engine) ListRunners() map[string]api.Runner {
-		k := g.Build.BuildKey()
-		uniq[k] = append(uniq[k], idx)
-	}
-
-	var finalSources []*api.UnpackedSources
-	if uniqcnt := len(uniq); uniqcnt == 1 {
-		finalSources = []*api.UnpackedSources{sources}
-	} else {
-		finalSources = make([]*api.UnpackedSources, uniqcnt)
-
-		for i := 0; i < uniqcnt; i++ {
-			dst := fmt.Sprintf("%s-%d", strings.TrimSuffix(sources.BaseDir, "/"), i)
-			if err := copy.Copy(sources.BaseDir, dst); err != nil {
-				return nil, fmt.Errorf("failed to create unique source directories for multiple build jobs: %w", err)
-			}
-			src := &api.UnpackedSources{
-				BaseDir: dst,
-				PlanDir: filepath.Join(dst, filepath.Base(sources.PlanDir)),
-			}
-			if sources.SDKDir != "" {
-				src.SDKDir = filepath.Join(dst, filepath.Base(sources.SDKDir))
-			}
-			if sources.ExtraDir != "" {
-				src.ExtraDir = filepath.Join(dst, filepath.Base(sources.ExtraDir))
-			}
-			finalSources[i] = src
-		}
-	}
-
-	// Trigger a build job for each unique build, and wait until all of them are
-	// done, mapping the build artifacts back to the original group positions in
-	// the response.
-	var cnt int
-	for key, idxs := range uniq {
-		idxs := idxs
-		key := key // capture
-
-		src := finalSources[cnt]
-		cnt++
-
-		errgrp.Go(func() (err error) {
-			// All groups are identical for the sake of building, so pick the first one.
-			grp := comp.Groups[idxs[0]]
-
-			// Pluck all IDs from the groups this build artifact is for.
-			grpids := make([]string, 0, len(idxs))
-			for _, idx := range idxs {
-				grpids = append(grpids, comp.Groups[idx].ID)
-			}
-
-			ow.Infow("performing build for groups", "plan", plan, "groups", grpids, "builder", builder)
-
-			deps := make(map[string]api.DependencyTarget, len(grp.Build.Dependencies))
-
-			for _, dep := range grp.Build.Dependencies {
-				deps[dep.Module] = api.DependencyTarget{
-					Target:  dep.Target,
-					Version: dep.Version,
-				}
-			}
-
-			in := &api.BuildInput{
-				BuildID:         uuid.New().String()[24:],
-				EnvConfig:       *e.envcfg,
-				TestPlan:        plan,
-				Selectors:       grp.Build.Selectors,
-				Dependencies:    deps,
-				BuildConfig:     obj,
-				UnpackedSources: src,
-			}
-
-			res, err := bm.Build(ctx, in, ow)
-			if err != nil {
-				ow.Infow("build failed", "plan", plan, "groups", grpids, "builder", builder, "error", err)
-				return err
-			}
-
-			res.BuilderID = bm.ID()
-
-			// no need for a mutex as the indices we access do not intersect
-			// across goroutines.
-			for _, idx := range uniq[key] {
-				ress[idx] = res
-			}
-
-			ow.Infow("build succeeded", "plan", plan, "groups", grpids, "builder", builder, "artifact", res.ArtifactPath)
-			return nil
-		})
-	}
-
-	// Wait until all goroutines are done. If any failed, return the error.
-	if err := errgrp.Wait(); err != nil {
-		return nil, err
-	}
-
-	return ress, nil
+func (e *Engine) QueueBuild(request *api.BuildRequest, sources *api.UnpackedSources) (string, error) {
+	id := uuid.New().String()
```

Comment:
I think it'd be best to stay with `uuid.New()[:8]` because these IDs are part of image names and we quickly hit docker tag name limits if we use a full-fledged UUID (36 chars). I think 8 chars random UUID is fine for Testground for the foreseeable future.